Feat/select atom

mhcoffin opened the pull request:
Let me know if I'm missing anything...

-mike

dai-shi commented on docs/api/utils.md:
```diff
@@ -244,75 +244,51 @@ const todoFamily = atomFamily(
+  anAtom: Atom<Value>,
+  selector: (v: Value) => Slice,
+  equalityFn: (a: Slice, b: Slice) => boolean = Object.is
+): Atom<Slice>
```
Do you think it's better to have TS in docs? You want to change `ts` in L251 then.

(btw, we should work on improving docs, and maybe it's better to make it consistent throughout the docs. would be for another PR.)

dai-shi commented on docs/api/utils.md:
```diff
@@ -244,75 +244,51 @@ const todoFamily = atomFamily(
 
-const stateAtom = atomWithReducer(initialState, reducer)
+// Original atom.
+const person = atom(defaultPerson)
```
```suggestion
const personAtom = atom(defaultPerson)
```

Lately, I realized that our convention is `(noun)+Atom` to represent an atom config.

(oth, `(verb)+Atom` is a utility function to do something with atoms.)

dai-shi commented on docs/api/utils.md:
```diff
@@ -244,75 +244,51 @@ const todoFamily = atomFamily(
 
-const selectCount = (state: State) => state.count
+// Tracks person.name; updated when person.name gets a new reference
+const name = selectAtom(defaultPerson, (person) => person.name)
```
```suggestion
const nameAtom = selectAtom(defaultPerson, (person) => person.name)
```

dai-shi commented on docs/api/utils.md:
```diff
@@ -244,75 +244,51 @@ const todoFamily = atomFamily(
-}
+// Tracks person.birth; updated when year, month, or day change
+import deepEqual from 'fast-deep-equal'
+const age = selectAtom(defaultPerson, (person) => person.birth, deeqEqual)
```
```suggestion
const ageAtom = selectAtom(defaultPerson, (person) => person.birth, deeqEqual)
```

Actually, for this specific example, shallowEqual is enough. maybe it's fine as is.

dai-shi commented on docs/api/utils.md:
```diff
@@ -244,75 +244,51 @@ const todoFamily = atomFamily(
 
-### Codesandbox
-
-https://codesandbox.io/s/react-typescript-forked-i4880
```
Would be nice if you can create one! You can use the codesandbox built by the ci in the PR.

dai-shi commented on tests/utils/selectAtom.test.tsx:
```diff
@@ -150,9 +146,11 @@ it('useSelector with scope', async () => {
-      bigAtom,
-      useCallback((value) => value.a, [])
+    const a = useAtomValue(
+      selectAtom(
+        bigAtom,
+        useCallback((value) => value.a, [])
+      )
```
```suggestion
      useMemo(() => selectAtom(
        bigAtom,
        (value) => value.a
      ), [bigAtom])
```

Inside component, we always need useMemo to make the derived atom stable. Let's do so in tests too.

dai-shi reviewed (commented):
Thanks for your work! Overall, looks good. Commented.

dai-shi commented:
@mhcoffin Do you have a chance to create a codesandbox example? (sorry if you already plan to work on it.)

mhcoffin commented:
Yes, working on it. Should have it done this morning (California).

On Thu, Feb 4, 2021 at 5:56 AM Daishi Kato <notifications@github.com> wrote:

> @mhcoffin <https://github.com/mhcoffin> Do you have a chance to create a
> codesandbox example? (sorry if you already plan to work on it.)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pmndrs/jotai/pull/290#issuecomment-773319544>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABEDBRVY5PFK7X7ZXFTEWXTS5KRRTANCNFSM4XBVI6SQ>
> .
>

mhcoffin commented:
On Wed, Feb 3, 2021 at 5:24 PM Daishi Kato <notifications@github.com> wrote:

> *@dai-shi* commented on this pull request.
>
> Thanks for your work! Overall, looks good. Commented.
> ------------------------------
>
> In docs/api/utils.md
> <https://github.com/pmndrs/jotai/pull/290#discussion_r569872671>:
>
> >  ```js
>
> -useSelector(anAtom, selector, equalityFn)
>
> +function selectAtom<Value, Slice>(
>
> +  anAtom: Atom<Value>,
>
> +  selector: (v: Value) => Slice,
>
> +  equalityFn: (a: Slice, b: Slice) => boolean = Object.is
>
> +): Atom<Slice>
>
>
> Do you think it's better to have TS in docs? You want to change ts in
> L251 then.
>
> (btw, we should work on improving docs, and maybe it's better to make it
> consistent throughout the docs. would be for another PR.)
>
I'll leave it consistent for now and volunteer to make another PR to switch
it all to ts.

> ------------------------------
>
> In docs/api/utils.md
> <https://github.com/pmndrs/jotai/pull/290#discussion_r569873992>:
>
> >  }
>
>
>
> -const stateAtom = atomWithReducer(initialState, reducer)
>
> +// Original atom.
>
> +const person = atom(defaultPerson)
>
>
> ⬇️ Suggested change
>
> -const person = atom(defaultPerson)
>
> +const personAtom = atom(defaultPerson)
>
>
> Lately, I realized that our convention is (noun)+Atom to represent an
> atom config.
>
> (oth, (verb)+Atom is a utility function to do something with atoms.)
>
Done.

> ------------------------------
>
> In docs/api/utils.md
> <https://github.com/pmndrs/jotai/pull/290#discussion_r569874085>:
>
> >
>
> -const selectCount = (state: State) => state.count
>
> +// Tracks person.name; updated when person.name gets a new reference
>
> +const name = selectAtom(defaultPerson, (person) => person.name)
>
>
> ⬇️ Suggested change
>
> -const name = selectAtom(defaultPerson, (person) => person.name)
>
> +const nameAtom = selectAtom(defaultPerson, (person) => person.name)
>
> Done.

> ------------------------------
>
> In docs/api/utils.md
> <https://github.com/pmndrs/jotai/pull/290#discussion_r569874443>:
>
> > -const TextBox = () => {
>
> -  const dispatch = useUpdateAtom(stateAtom)
>
> -  const text = useSelector(stateAtom, selectText)
>
> -  return (
>
> -    <div>
>
> -      {text}{' '}
>
> -      <input
>
> -        value={text}
>
> -        onChange={(e) => dispatch({ type: 'SET_TEXT', text: e.target.value })}
>
> -      />
>
> -    </div>
>
> -  )
>
> -}
>
> +// Tracks person.birth; updated when year, month, or day change
>
> +import deepEqual from 'fast-deep-equal'
>
> +const age = selectAtom(defaultPerson, (person) => person.birth, deeqEqual)
>
>
> ⬇️ Suggested change
>
> -const age = selectAtom(defaultPerson, (person) => person.birth, deeqEqual)
>
> +const ageAtom = selectAtom(defaultPerson, (person) => person.birth, deeqEqual)
>
>
> Actually, for this specific example, shallowEqual is enough. maybe it's
> fine as is.
>
I made the birth struct a little deeper to emphasize the comparison
function. Also renamed to birthAtom to match the name of the field.

> ------------------------------
>
> In docs/api/utils.md
> <https://github.com/pmndrs/jotai/pull/290#discussion_r569874791>:
>
> >  ```
>
>
>
> -### Codesandbox
>
> -
>
> -https://codesandbox.io/s/react-typescript-forked-i4880
>
> Would be nice if you can create one! You can use the codesandbox built by
> the ci in the PR.
>
Done.

> ------------------------------
>
> In tests/utils/selectAtom.test.tsx
> <https://github.com/pmndrs/jotai/pull/290#discussion_r569876703>:
>
> > +      selectAtom(
>
> +        bigAtom,
>
> +        useCallback((value) => value.a, [])
>
> +      )
>
>
> ⬇️ Suggested change
>
> -      selectAtom(
>
> -        bigAtom,
>
> -        useCallback((value) => value.a, [])
>
> -      )
>
> +      useMemo(() => selectAtom(
>
> +        bigAtom,
>
> +        (value) => value.a
>
> +      ), [bigAtom])
>
>
> Inside component, we always need useMemo to make the derived atom stable.
> Let's do so in tests too.
>
Done. (But bigAtom is an outer-scope const, so shouldn't be in deps.)

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pmndrs/jotai/pull/290#pullrequestreview-582954648>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABEDBRQRVXYGEZV3QAEIAU3S5HZM3ANCNFSM4XBVI6SQ>
> .
>

dai-shi reviewed (approved):
Looks great! Thanks so much for your work.

mhcoffin commented:
No problem.  Glad to help.

-mike

On Thu, Feb 4, 2021 at 3:19 PM Daishi Kato <notifications@github.com> wrote:

> *@dai-shi* approved this pull request.
>
> Looks great! Thanks so much for your work.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pmndrs/jotai/pull/290#pullrequestreview-583897314>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABEDBRXZBLV4LDRTTF572WLS5MTPVANCNFSM4XBVI6SQ>
> .
>

dai-shi commented:
btw we renamed #264 splitAtom from sliceAtom. so, this can be technically changed. maybe confusing anyway. `selectAtom` is good to go.
